src/write_events.py

```diff
@@ -0,0 +1,40 @@
+import pandas as pd
+from ata_db_models.models import Event
+from sqlalchemy.dialects.postgresql import insert
+from sqlalchemy.orm import sessionmaker
+
+from src.helpers.logging import logging
+
+logger = logging.getLogger(__name__)
+
+
+def write_events(df: pd.DataFrame, Session: sessionmaker) -> int:
+    """
+    Writes preprocessed events to database.
+
+    This function accepts a `sessionmaker`, which is a factory for session
+    objects, given an engine. A `sessionmaker` can be created like so:
+    >>> Session = sessionmaker(engine)
+    """
+    data = df.to_dict(orient="records")
+
+    # Create statement to bulk-insert event rows
+    # Insert.on_conflict_do_nothing skips through events whose [event_id, site_name]
+    # composite key already exists in the DB
+    statement = insert(Event).values(data).on_conflict_do_nothing(index_elements=[Event.site_name, Event.event_id])
+
+    # Wrap execution within a begin-commit-rollback block in the form of two
+    # context managers (see: https://docs.sqlalchemy.org/en/14/orm/session_basics.html#framing-out-a-begin-commit-rollback-block)
+    with Session() as session, session.begin():
```

Comment:
🔧 Don't call `begin`, [call `commit` or set the session to autocommit](https://docs.sqlalchemy.org/en/14/orm/session_basics.html#basics-of-using-a-session).
```suggestion
    with Session() as session:
```

You can call `session.commit()` at the end of the `with` block, or set the sessionmaker to generate sessions that autocommit.

Reply:
How about just `with session_factory.begin() as session`? The returned Session object [autocommits](https://docs.sqlalchemy.org/en/14/orm/session_api.html#sqlalchemy.orm.sessionmaker.begin) with the added benefit of being able to rollback in the event of an exception (assuming we want to). From [this page](https://docs.sqlalchemy.org/en/14/orm/session_transaction.html#managing-transactions):

> At the end of the above context, assuming no exceptions were raised, any pending objects will be flushed to the database and the database transaction will be committed. If an exception was raised within the above block, then the transaction would be rolled back
>
> [...]
>
>[sessionmaker](https://docs.sqlalchemy.org/en/14/orm/session_api.html#sqlalchemy.orm.sessionmaker) itself includes a [sessionmaker.begin()](https://docs.sqlalchemy.org/en/14/orm/session_api.html#sqlalchemy.orm.sessionmaker.begin) method to allow both operations to take place at once:
>```python
> with Session.begin() as session:
>   session.add(some_object)
> ```

Reply:
Sure thing. Man they have way too many ways to do things

Reply:
Hehe done

Reply:
Passed tests but mypy's complaining. Working on it now.

Reply:
Alrighty, should be good.